lib/output_descriptor.ex

```diff
@@ -0,0 +1,527 @@
+            {:ok, point} -> {:ok, point, rest}
+            {:error, reason} -> {:error, reason}
+          end
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+    else
+      {:error, "Invalid public key format"}
+    end
+  end
+
+  defp parse_multi_descriptor(str, sorted?) do
+    # Parse threshold
+    case Integer.parse(str) do
+      {threshold, "," <> rest} ->
+        # Parse keys
+        case parse_multi_keys(rest, []) do
+          {:ok, keys, remaining} ->
+            case parse_closing_bracket(remaining) do
+              {:ok, rest_with_checksum} ->
+                checksum = extract_checksum(rest_with_checksum)
+                type = if sorted?, do: :sortedmulti, else: :multi
+
+                {:ok,
+                 %__MODULE__{
+                   type: type,
+                   threshold: threshold,
+                   keys: keys,
+                   checksum: checksum
+                 }}
+
+              {:error, reason} ->
+                {:error, reason}
+            end
+
+          {:error, reason} ->
+            {:error, reason}
+        end
+
+      _ ->
+        {:error, "Invalid multi descriptor format"}
+    end
+  end
+
+  defp parse_multi_keys(str, acc) do
+    case parse_key_or_descriptor(str) do
+      {:error, reason} ->
+        {:error, reason}
+
+      {{:key, key}, ")" <> _rest} ->
+        {:ok, Enum.reverse([key | acc]), ")" <> str}
+
+      {{:key, key}, "," <> rest} ->
+        parse_multi_keys(rest, [key | acc])
+
+      _ ->
+        {:error, "Invalid key format in multi descriptor"}
+    end
+  end
+
+  defp parse_until_closing_bracket(str) do
+    case String.split(str, ")", parts: 2) do
+      [content, rest] -> {content, ")" <> rest}
+      _ -> {:error, "Missing closing bracket"}
+    end
+  end
+
+  defp parse_closing_bracket(")" <> rest), do: {:ok, rest}
+  defp parse_closing_bracket(_), do: {:error, "Missing closing bracket"}
+
+  defp parse_until_matching_bracket(<<>>, _depth, _start_pos, _acc),
+    do: {:error, "Unmatched brackets"}
+
+  defp parse_until_matching_bracket(<<"(", rest::binary>>, depth, start_pos, acc) do
+    parse_until_matching_bracket(rest, depth + 1, start_pos, acc <> "(")
+  end
+
+  defp parse_until_matching_bracket(<<")", rest::binary>>, 1, _start_pos, acc) do
+    {:ok, acc, ")" <> rest}
+  end
+
+  defp parse_until_matching_bracket(<<")", rest::binary>>, depth, start_pos, acc)
+       when depth > 1 do
+    parse_until_matching_bracket(rest, depth - 1, start_pos, acc <> ")")
+  end
+
+  defp parse_until_matching_bracket(<<c::binary-size(1), rest::binary>>, depth, start_pos, acc) do
+    parse_until_matching_bracket(rest, depth, start_pos, acc <> c)
+  end
+
+  defp extract_checksum(""), do: nil
+  defp extract_checksum("#" <> checksum), do: checksum
+
+  @doc """
+  Converts an output descriptor to a script.
+  """
+  @spec to_script(t()) :: Script.t()
+  def to_script(%__MODULE__{type: :pk, key: key}) do
```

Comment:
i think in each of these, you will need to add a `when key != nil` otherwise `key: key` will always match.

Reply:
also, there's no handling of cases where an xpub is provided, meaning this function would crash. we should add sth like `def to_script(_), do: {:error, "..."}` or, even better, allow the caller to pass in an `opts` which contains a derivation path, or other necessary args to derive a valid script.